src/zeit/content/cp/automatic.py

```diff
@@ -67,116 +54,170 @@ def values(self):
         result = []
         try:
-            for item in zeit.find.search.search(
-                    query, sort_order=sort_order,
-                    start=self._v_retrieved_content,
-                    rows=self.count_to_replace_duplicates):
-                content = zeit.cms.interfaces.ICMSContent(
-                    item['uniqueId'], None)
+            solr = zope.component.getUtility(zeit.solr.interfaces.ISolr)
+            response = solr.search(
+                query, sort=sort_order,
+                start=self.start,
+                rows=self.rows,
+                fl=self.FIELDS,
+                fq=self.filter_query)
+            self.total_hits = response.hits
+            for item in response:
+                content = self._resolve(item)
                 if content is not None:
                     result.append(content)
         except:
             log.warning('Error during solr query %r for %s',
-                        query, self.uniqueId, exc_info=True)
+                        query, self.context.uniqueId, exc_info=True)
         return result
 
-    def _query_centerpage(self):
-        teasered = zeit.content.cp.interfaces.ITeaseredContent(
-            self.referenced_cp, iter([]))
-        result = []
-        for i in range(
-                self._v_retrieved_content + self.count_to_replace_duplicates):
-            try:
-                content = teasered.next()
-            except StopIteration:
-                # We've exhausted the available content.
-                break
-            if i >= self._v_retrieved_content:
-                result.append(content)
-        return result
+    def _resolve(self, solr_result):
+        return zeit.cms.interfaces.ICMSContent(solr_result['uniqueId'], None)
 
-    def _extract_newest(self, content, predicate=None):
-        """Remove the first object from the content list for which predicate
-        returns True; thus, the default predicate means: no filtering.
-        """
-        result = None
-        pop = []
-        cp = zeit.content.cp.interfaces.ICenterPage(self)
-        for i, item in enumerate(content):
-            if predicate is None or predicate(item):
-                pop.append(item)
-                if self.hide_dupes and (
-                        cp.is_teaser_present_above(self.context, item)
-                        or cp.is_teaser_manual_below(self.context, item)):
-                    continue
-                result = item
-                break
-        for item in pop:
-            content.remove(item)
-        if result is None and not content:
-            # We've exhausted all available content due to duplicates, so we
-            # need to retrieve some more.
-            more_content = self._retrieve_content()
-            if more_content:
-                content[:] = more_content
-                return self._extract_newest(content, predicate)
-        return result
+    @property
+    def filter_query(self):
```

Comment:
Der Methode würde eine Beschreibung gut tun. So viele "not", das ist schwer zu lesen.